velox/exec/Spill.h

```diff
@@ -0,0 +1,350 @@
+  }
+
+  // Prepares 'this' for reading. Positions the read at the first row of
+  // content.
+  void startRead();
+
+  // Returns the file size in bytes.
+  uint64_t size() const override {
+    if (output_) {
+      return output_->size();
+    }
+    return fileSize_;
+  }
+
+  // True if no more SpillFileRows can be read from 'this'.
+  bool atEnd() const override {
+    return index_ >= numRowsInVector_ && input_->atEnd();
+  }
+
+  // Sets 'result' to refer to the next row of content of 'this'.
+  void read(RowVector& result);
+
+ protected:
+  void nextBatch() override {
+    VectorStreamGroup::read(input_.get(), &pool_, type_, &rowVector_);
+    index_ = 0;
+    numRowsInVector_ = rowVector_->size();
+  }
+
+ private:
+  std::string path_;
+  // Byte size of the backing file. Set when finishing writing.
+  uint64_t fileSize_ = 0;
+  std::unique_ptr<WriteFile> output_;
+  std::unique_ptr<SpillInput> input_;
+};
+
+// Describes a bit range inside a 64 bit hash number for use in
+// partitioning data over multiple sets of spill files.
+struct HashBitRange {
+  // Low bit number of hash number bit range
+  uint8_t begin;
+  // Bit number of first bit above the hash number bit range.
+  uint8_t end;
+};
+
+// Sequence of files for one partition of the spilled data. If data is
+// sorted, each file is sorted. The globally sorted order is produced
+// by merging the constituent files.
+class SpillFileList {
+ public:
+  // Constructs a set of spill files. 'type' is a RowType describing the
+  // content. 'path' is a file path prefix. 'targetBatchSize is the target size
+  // of a single RowVector in rows. 'targetFileSize' is the target byte size of
+  // a single file in the file set. 'pool' and 'mappedMemory' are used for
+  // buffering and constructing the result data read from 'this'.
+  SpillFileList(
+      RowTypePtr type,
+      const std::string path,
+      uint64_t targetBatchSize,
+      uint64_t targetFileSize,
+      memory::MemoryPool& pool,
+      memory::MappedMemory& mappedMemory)
+      : type_(type),
+        path_(path),
+        targetBatchSize_(targetBatchSize),
+        targetFileSize_(targetFileSize),
+        pool_(pool),
+        mappedMemory_(mappedMemory) {}
+
+  // Adds 'rows' for the positions in 'indices' into 'this'. The indices
+  // must produce a view where the rows are sorted if sorting is desired.
+  // Consecutive calls must have sorted data so that the first row of the
+  // next call is not less than the last row of the previous call.
+  void write(RowVectorPtr rows, const folly::Range<IndexRange*> indices);
+  std::vector<std::unique_ptr<SpillFile>> files() {
+    return std::move(files_);
+  }
+
+  // Closes one file of output. Subsequent calls to 'write' start a
+  // different sorted run and must be ordered between themselves but not
+  // with respect to calls before flush(). If 'close' is true, write must
+  // not be called and the file set can only be read.
+  void flush(bool close);
```

Comment:
Having both targetFileSize and flush(bool close) method in the API is confusing. targetFileSize suggests that the class decides on its own when to close current file and start a new one, but flush(bool close) method suggests that the caller is expected to call flush multiple times to trigger actual closing of one file and start of the new. It would be nice to make the API less confusing and document how targetFileSize is actually used.